pyhap/accessory_driver.py

```diff
@@ -96,6 +99,32 @@ class HAP_CONSTANTS:
     SERVICE_COMMUNICATION_FAILURE = -70402
 
 
+class AIOThread(threading.Thread):
+    """
+    TODO: temporary convenience class. Create an event loop and event object for
+    controlling the Accessory.run method.
+    """
+
+    def __init__(self, run_method):
+        """
+        """
+        self.loop = asyncio.new_event_loop()
+        self.stop_event = asyncio.Event()
+        self.task = self.loop.create_task(run_method(self.stop_event, self.loop))
+        super(AIOThread, self).__init__(target=self.loop.run_until_complete,
+                                        args=(self.task,))
+
+    def run(self):
+        try:
+            super(AIOThread, self).run()
+        except CancelledError:
+            self.loop.stop()
+            self.loop.close()
+            logger.info("Sucessfully stopped accessory event loop.")
+
+    def stop(self):
+        self.loop.call_soon_threadsafe(self.task.cancel)
+
```

Comment:
I don't think this class is safe. It looks like you are using a single event loop from multiple threads here. I'm 99% sure you shouldn't do that. (I've had exceptions thrown as a result).

Instead the event loop should be setup after thread creation like this:

```python
class AIOThread(threading.Thread):
    """
    TODO: temporary convenience class. Create an event loop and event object for
    controlling the Accessory.run method.
    """

    def __init__(self, run_method):
        """
        """
        self.run_method = run_method
        super().__init__()

    def run(self):
        self.loop = asyncio.new_event_loop()
        asyncio.set_event_loop(self.loop)

        self.stop_event = asyncio.Event()
        self.task = self.loop.create_task(self.run_method(self.stop_event, self.loop))

        self.loop.run_until_complete(self.task)

    async def safe_shutdown(self):
        try:
            self.stop_event.set()
            await asyncio.sleep(5)
            if not self.task.done():
                self.task.cancel()
                await asyncio.sleep(5)
            self.loop.stop()
        finally:
            self.loop.close()
            logger.info("Sucessfully stopped accessory event loop.")

    def stop(self):
        self.loop.call_soon_threadsafe(self.safe_shutdown)
```